File: src/main/java/org/dukcode/ps/codetree/trail02/chapter04/lesson03/IntroMaximumOverlappedSegments.java
package org.dukcode.ps.codetree.trail02.chapter04.lesson03;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-maximum-overlapped-segments/description">LINK</a>
 */
public class IntroMaximumOverlappedSegments {

  private static final int RANGE = 100;

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int[] lines = new int[2 * RANGE + 1];

    int ans = 0;
    int n = Integer.parseInt(br.readLine());
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      int frIn = Integer.parseInt(st.nextToken());
      int toEx = Integer.parseInt(st.nextToken());

      for (int idx = frIn; idx < toEx; idx++) {
        lines[idx + RANGE]++;
        ans = Math.max(ans, lines[idx + RANGE]);
      }
    }

    bw.write(String.valueOf(ans));

    br.close();
    bw.close();

  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter04/lesson03/IntroMaximumOverlappedSegments.kt
package org.dukcode.ps.codetree.trail02.chapter04.lesson03.introMaximumOverlappedSegments

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

private const val MX = 100

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-maximum-overlapped-segments/description">LINK</a>
 */
fun main() {
    val n = br.readLine().toInt()

    val arr = IntArray(2 * MX + 1)
    var mx = 0
    repeat(n) {
        val (frIn, toIn) = parseInts()

        for (idx in frIn until toIn) {
            arr[idx + MX]++
            mx = maxOf(mx, arr[idx + MX])
        }
    }

    bw.write(mx.toString())

    bw.flush()
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
